.circleci/config.yml

```diff
@@ -1,74 +1,115 @@
-version: 2
-jobs:
-  build:
-    environment:
-      CIRCLE_ARTIFACTS:    /tmp/circleci-artifacts
-      CIRCLE_TEST_REPORTS: /tmp/circleci-test-results
+version: 2.1
+
+executors:
+  openjdk_executor:
     docker:
       - image: circleci/openjdk@sha256:012843a8f91da1ea5e22e142bb5cb8c999a7457e03b8c79ab45d5fcb68043383
-    steps:
-    - checkout
-    - run:
-        name: Create output directories
-        command: mkdir -p $CIRCLE_ARTIFACTS $CIRCLE_TEST_REPORTS $CIRCLE_TEST_REPORTS/junit/
-
-    - run:
-        name: Log Java version
-        command: java -version
-
-    # Dependency caching
-    - restore_cache:
-        keys:
-          # when lock file changes, use increasingly general patterns to restore cache
-          - gradle-repo-v5-{{ .Branch }}-{{ checksum "gradle.properties" }}
-          - gradle-repo-v5-{{ .Branch }}-
-          - gradle-repo-v5-
-    - run:
-        name: Download uncached dependencies
-        command: ./gradlew dependencies
-        no_output_timeout: 1h
-
-    - save_cache:
-        paths:
-          - ~/.gradle
-        key: gradle-repo-v5-{{ .Branch }}-{{ checksum "gradle.properties" }}
-
-    # Compile
-    - run:
-        name: Compile
-        command: xvfb-run ./gradlew -PbuildNumber=$CIRCLE_BUILD_NUM -PbuildAccess=opensource -PbuildTag=$CIRCLE_TAG --refresh-dependencies --info --stacktrace clean assemble
+    environment:
+      _JAVA_OPTIONS: "-Xmx3g"
+      GRADLE_OPTS: "-Dorg.gradle.daemon=false"
 
-    # Test
-    - run:
-        name: Run tests
-        command: xvfb-run ./gradlew -PbuildAccess=opensource -PbuildNumber=$CIRCLE_BUILD_NUM -PbuildTag=$CIRCLE_TAG --info --stacktrace test
+jobs:
+  checkout_run_test:
+    executor: openjdk_executor
+    description: "Perform all setup steps"
+    steps:
+        - checkout
+        - run:
+            name: Create output directories
+            command: "mkdir -vp ~/circleci-artifacts ~/circleci-test-results ~/circleci-test-results/junit/"
+        
+        - restore_cache:
+            keys:
+                - gradle-repo-v5-{{ .Branch }}-{{ checksum "gradle.properties" }}
+                - gradle-repo-v5-{{ .Branch }}-
+                - gradle-repo-v5-
+        - run:
+            name: Download uncached dependencies
+            command: ./gradlew dependencies
+            no_output_timeout: 1h
+        
+        - save_cache:
+            paths:
+                - ~/.gradle
+            key: gradle-repo-v5-{{ .Branch }}-{{ checksum "gradle.properties" }}
+        
+        - run:
+            name: Compile
+            command: xvfb-run ./gradlew -PbuildNumber=$CIRCLE_BUILD_NUM -PbuildAccess=opensource -PbuildTag=$CIRCLE_TAG --refresh-dependencies --info --stacktrace clean assemble
```

Comment:
Compiling shouldn't require a desktop environment. Let's confirm that and remove `xvfb-run`.